### PR TITLE
refactoring pfio error trapping

### PR DESCRIPTION
**Pull request by weiyuan-jiang**
add more error trapping chain to pFIO
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Trivial change (affects only documentation or cleanup)

## Checklist:

- [ ] I have tested this change with a run of GEOSgcm (if non-trivial)
- [x] I have added one of the required labels (0 diff, 0 diff trivial, 0 diff structural, non 0-diff)
- [x] I have updated the CHANGELOG.md accordingly following the style of [Keep a Changelog](https://keepachangelog.com/en/1.0.0/#how)

### CODE REVIEW

File: pfio/AbstractServer.F90
Comment (tclune) [inline pfio/AbstractServer.F90:168]: status is not being checked here?

File: pfio/AbstractServer.F90
Comment (tclune) [inline pfio/AbstractServer.F90:133]: `ierror` is the name of the argument in MPI.   `status` does not buy us much unless it somehow allows the use of a better macro, which does not seem to be the case.

File: pfio/AbstractServer.F90
Comment (weiyuan-jiang) [reply pfio/AbstractServer.F90:168]: Do we always check Mpi's error? I skip a lot

File: griddedio/FieldBundleRead.F90
Comment (tclune) [inline griddedio/FieldBundleRead.F90:61]: Please remember to start using `__RC__` for these cases.   It helps to unclutter the code.  Not worth a separate commit in just this instance, but all future changes in MAPL should try to follow this best practice.

File: pfio/AbstractServer.F90
Comment (weiyuan-jiang) [reply pfio/AbstractServer.F90:133]: Do you mean if we don't want to check it,  keep it as ierror ?

**Review by tclune — CHANGES_REQUESTED** [review]
1. Does not address the `get_var()` and `put_var()` error handling.   
2. Exposed missing checks on error code from MPI calls in the same layer.

Also various minor comments about style.

File: pfio/AbstractServer.F90
Comment (tclune) [reply pfio/AbstractServer.F90:133]: Either way we must check the return code.

So either leave it as `ierror` and use `_VERIFY(ierror)`

Or call it `status` and use  `__STAT__` macro (akin to `__RC__` macro).

One could debate the merits of introducing an MPI specific macro `_IERR` that does the same thing with `ierror` but that's probably not worth the trouble.

File: base/ServerManager.F90
Comment (mathomp4) [inline base/ServerManager.F90:164]: These might as well go to `__RC__` as well. I think there are a few others here.

File: pfio/CMakeLists.txt
Comment (mathomp4) [inline pfio/CMakeLists.txt:95]: There is ESMF in pFIO? That is surprising. Are we using some utility from it?

File: pfio/MultiGroupServer.F90
Comment (mathomp4) [inline pfio/MultiGroupServer.F90:247]: It looks like `report_profiler()` might have an RC code...but maybe it's not worth checking?

File: pfio/AbstractServer.F90
Comment (weiyuan-jiang) [reply pfio/AbstractServer.F90]: changed

File: pfio/AbstractServer.F90
Comment (weiyuan-jiang) [reply pfio/AbstractServer.F90]: changed

File: griddedio/FieldBundleRead.F90
Comment (weiyuan-jiang) [reply griddedio/FieldBundleRead.F90]: I have changed them to __RC__ .  Now pFIO depends on esmf, which should be fine

File: base/ServerManager.F90
Comment (tclune) [inline base/ServerManager.F90:263]: `_VERIFY` is not needed with `__RC__`
```suggestion
```

File: base/ServerManager.F90
Comment (tclune) [inline base/ServerManager.F90:232]:
```suggestion
              allocate(this%o_server, source = MpiServer(this%split_comm%get_subcommunicator(), s_name, with_profiler=with_profiler, __RC__))
```

File: base/ServerManager.F90
Comment (weiyuan-jiang) [reply base/ServerManager.F90:164]: There are a few others, but not this one.

File: pfio/CMakeLists.txt
Comment (weiyuan-jiang) [reply pfio/CMakeLists.txt:95]: __RC__ need "MAPL_Exceptions.h" which includes "ESMC_ReturnCodes.h"

File: base/ServerManager.F90
Comment (mathomp4) [reply base/ServerManager.F90:164]: Ohhh. Duh. I didn't see this was inside an `allocate()`. I just saw `rc=status` and my brain went into overdrive. 😄

Comment (tclune) [reply base/ServerManager.F90:164]: Ah - we have to be careful with these.   The expression with `rc=status` is itself inside an `ALLOCATE` statement, so there are potentially two different failures.    Ordinarily I prefer not to worry about insufficient memory when allocating, but in these cases, if the expression for SOURCE fails, then the ALLOCATE fails for more immediate reasons.    So we should check both.     

Further, we need to check the expression result first as it is the more meaningful one.    So we need two different status's and can't use the macros at all.
```suggestion
           allocate(this%i_server, source = MpiServer(client_comm, 'i_server'//trim(i_to_string(1)),profiler_name=profiler_name, &
                                                      with_profiler=with_profiler, rc=status), stat=status_alloc)
           _VERIFY(status)
           _VERIFY(status_alloc)
```
You'll need to add the new integer `status_alloc` in the top of the procedure.  (Not wed to this name.)

File: base/ServerManager.F90
Comment (tclune) [inline base/ServerManager.F90:174]: Same issue as above.

File: base/ServerManager.F90
Comment (tclune) [inline base/ServerManager.F90:228]: Same issue as above.

File: base/ServerManager.F90
Comment (tclune) [inline base/ServerManager.F90:232]: And again.

File: pfio/CMakeLists.txt
Comment (tclune) [reply pfio/CMakeLists.txt:95]: I agree - but maybe this got relaxed somewhere.   Really prefer for pfio to be independent of ESMF.

File: pfio/MultiGroupServer.F90
Comment (tclune) [reply pfio/MultiGroupServer.F90:247]: If it's got one, it should be checked.   Just to save the sanity of the next guy. 

Or, if we can prove that there is no path that generates an error, the RC code should be eliminated from the routine.

**Review by tclune — CHANGES_REQUESTED** [review]
Wow - this one really snowballed on us ...

**Comment by mathomp4** [comment]
I keep hoping to test this soon, but...yeesh, I can't get a few nodes at NCCS. Well, I can't get Cascade Lake...but:
```
Resources in idle state:

     Node Type   Nodes   Cores
  ------------   -----   -----
       Haswell       1      28
       Skylake     506   20240
  Cascade Lake       4     192
      Rome/GPU       8     384
```
I guess it's Skylake day for Matt!

File: base/ServerManager.F90
Comment (mathomp4) [reply base/ServerManager.F90:164]: I'm unresolving this. I think I resolved it just about when @tclune commented, so it's possible @weiyuan-jiang never saw this!.

But as @tclune says, it's probably good to test both returns here!

**Review by tclune — DISMISSED** [review]
Huzzah!

**Comment by mathomp4** [comment]
First: Nuts. Looks like my `if fail, re-run install` works, but CircleCI lists it as a failure because one step failed. Grr. Okay I'll need to think on that.

Second, @weiyuan-jiang an error still isn't being trapped. I just ran this branch and:

```
...
 AGCM Date: 2000/04/14  Time: 23:30:00  Throughput(days/day)[Avg Tot Run]:   1572.5   2363.0   2414.2  TimeRemaining(Est) 000:00:49   22.8% :  10.7% Mem Comm:Used
 AGCM Date: 2000/04/14  Time: 23:45:00  Throughput(days/day)[Avg Tot Run]:   1630.5   2586.1   2651.2  TimeRemaining(Est) 000:00:46   22.8% :  10.6% Mem Comm:Used

 Writing:    483 Slices to File:  weiyuanmapl-gcm-2022Jan11-1day-c24-ChunkingFailure.geosgcm_prog.20000415_0000z.nc4
 Writing:    153 Slices to File:  weiyuanmapl-gcm-2022Jan11-1day-c24-ChunkingFailure.geosgcm_surf.20000414_2230z.nc4
 Writing:      5 Slices to File:  weiyuanmapl-gcm-2022Jan11-1day-c24-ChunkingFailure.geosgcm_budi.20000415_0000z.nc4
 Writing:     21 Slices to File:  weiyuanmapl-gcm-2022Jan11-1day-c24-ChunkingFailure.geosgcm_buda.20000414_2230z.nc4
pe=00000 FAIL at line=00701    NetCDF4_FileFormatter.F90                <status=-127>
pe=00000 FAIL at line=00297    NetCDF4_FileFormatter.F90                <status=-127>
pe=00000 FAIL at line=00063    HistoryCollection.F90                    <status=-127>
pe=00000 FAIL at line=00792    ServerThread.F90                         <status=-127>
pe=00000 FAIL at line=00029    NetCDF4_put_var.H                        <status=-49>
pe=00000 FAIL at line=00836    ServerThread.F90                         <status=-49>
pe=00000 FAIL at line=00029    NetCDF4_put_var.H                        <status=-49>
pe=00000 FAIL at line=00836    ServerThread.F90                         <status=-49>
pe=00000 FAIL at line=00029    NetCDF4_put_var.H                        <status=-49>
pe=00000 FAIL at line=00836    ServerThread.F90                         <status=-49>
pe=00000 FAIL at line=00029    NetCDF4_put_var.H                        <status=-49>
pe=00000 FAIL at line=00836    ServerThread.F90                         <status=-49>
pe=00000 FAIL at line=00029    NetCDF4_put_var.H                        <status=-49>
pe=00000 FAIL at line=00836    ServerThread.F90                         <status=-49>
pe=00000 FAIL at line=00029    NetCDF4_put_var.H                        <status=-49>
pe=00000 FAIL at line=00836    ServerThread.F90                         <status=-49>
pe=00000 FAIL at line=00029    NetCDF4_put_var.H                        <status=-49>
pe=00000 FAIL at line=00836    ServerThread.F90                         <status=-49>
pe=00000 FAIL at line=00029    NetCDF4_put_var.H                        <status=-49>
pe=00000 FAIL at line=00836    ServerThread.F90                         <status=-49>
pe=00000 FAIL at line=00029    NetCDF4_put_var.H                        <status=-49>
pe=00000 FAIL at line=00836    ServerThread.F90                         <status=-49>
pe=00000 FAIL at line=00029    NetCDF4_put_var.H                        <status=-49>
pe=00000 FAIL at line=00836    ServerThread.F90                         <status=-49>
pe=00000 FAIL at line=00029    NetCDF4_put_var.H                        <status=-49>
pe=00000 FAIL at line=00836    ServerThread.F90                         <status=-49>
pe=00000 FAIL at line=00029    NetCDF4_put_var.H                        <status=-49>
pe=00000 FAIL at line=00836    ServerThread.F90                         <status=-49>
pe=00000 FAIL at line=00029    NetCDF4_put_var.H                        <status=-49>
pe=00000 FAIL at line=00836    ServerThread.F90                         <status=-49>

 AGCM Date: 2000/04/15  Time: 00:00:00  Throughput(days/day)[Avg Tot Run]:   1405.2    557.8   1902.1  TimeRemaining(Est) 000:00:53   22.8% :  10.7% Mem Comm:Used
        EXTDATA: INFO: Updating R bracket for BC_BIOMASS
        EXTDATA: INFO:  ... file processed: ExtData/PIESA/sfc/QFED/v2.4r6/Y2000/M04/qfed2.emis_bc.005.20000416.nc4
        EXTDATA: INFO: Updating R bracket for CO_BIOMASS
        EXTDATA: INFO:  ... file processed: ExtData/PIESA/sfc/QFED/v2.4r6/Y2000/M04/qfed2.emis_co.005.20000416.nc4
        EXTDATA: INFO: Updating R bracket for CO_ISOP
        EXTDATA: INFO:  ... file processed: ExtData/AeroCom/sfc/co_biogenic.x288_y181_t12.2006.nc
        EXTDATA: INFO: Updating R bracket for CO_NVOC
        EXTDATA: INFO:  ... file processed: ExtData/AeroCom/sfc/co_biogenic.x288_y181_t12.2006.nc
        EXTDATA: INFO: Updating R bracket for CO_TERP
        EXTDATA: INFO:  ... file processed: ExtData/AeroCom/sfc/co_biogenic.x288_y181_t12.2006.nc
        EXTDATA: INFO: Updating R bracket for EMI_NH3_BB
        EXTDATA: INFO:  ... file processed: ExtData/PIESA/sfc/QFED/v2.4r6/Y2000/M04/qfed2.emis_nh3.005.20000416.nc4
        EXTDATA: INFO: Updating R bracket for OC_BIOMASS
        EXTDATA: INFO:  ... file processed: ExtData/PIESA/sfc/QFED/v2.4r6/Y2000/M04/qfed2.emis_oc.005.20000416.nc4
        EXTDATA: INFO: Updating R bracket for SU_BIOMASS
        EXTDATA: INFO:  ... file processed: ExtData/PIESA/sfc/QFED/v2.4r6/Y2000/M04/qfed2.emis_so2.005.20000416.nc4

 PCHEM::Run: Indices selected from climatology are          255         256


 Updated CO2 in solar for year/day 2000/106 is  0.36855E-03

 AGCM Date: 2000/04/15  Time: 00:15:00  Throughput(days/day)[Avg Tot Run]:   1412.2   1503.3   2390.1  TimeRemaining(Est) 000:00:52   22.8% :  10.7% Mem Comm:Used
 AGCM Date: 2000/04/15  Time: 00:30:00  Throughput(days/day)[Avg Tot Run]:   1455.1   2411.0   2465.6  TimeRemaining(Est) 000:00:50   22.8% :  10.7% Mem Comm:Used
 AGCM Date: 2000/04/15  Time: 00:45:00  Throughput(days/day)[Avg Tot Run]:   1498.5   2577.1   2642.1  TimeRemaining(Est) 000:00:48   22.8% :  10.7% Mem Comm:Used
...
```

**Comment by mathomp4** [comment]
My current guess is this might be due to untrapped exceptions from `put_DataToFile`. I see some like here:

https://github.com/GEOS-ESM/MAPL/blob/2aa24510f371efddd9435ea0227425a335678120/pfio/BaseServer.F90#L138

But I have no idea which one I'm triggering. I see them in a few places:
```
$ rg 'call.*datatofile'
pfio/MultiGroupServer.F90
799:               call thread_ptr%put_dataToFile(q, address)

pfio/BaseServer.F90
138:              call threadPtr%put_DataToFile(q,offset_address)

pfio/MultiCommServer.F90
189:            call this%put_dataToFile(__RC__)
495:                  call threadPtr%put_DataToFile(q,offset_address)

pfio/ServerThread.F90
982:      call this%containing_server%put_dataToFile(rc=status)
1019:            call this%put_DataToFile(q,handle%data_reference%base_address,rc=status)
```

I'll tried adding some and now things are...interesting. The model "halts":
```
 Writing:      5 Slices to File:  weiyuanmapl-gcm-2022Jan11-1day-c24-ChunkingFailure.geosgcm_budi.20000415_0000z.nc4
 Writing:     21 Slices to File:  weiyuanmapl-gcm-2022Jan11-1day-c24-ChunkingFailure.geosgcm_buda.20000414_2230z.nc4
pe=00000 FAIL at line=00701    NetCDF4_FileFormatter.F90                <status=-127>
pe=00000 FAIL at line=00297    NetCDF4_FileFormatter.F90                <status=-127>
pe=00000 FAIL at line=00063    HistoryCollection.F90                    <status=-127>
pe=00000 FAIL at line=00795    ServerThread.F90                         <status=-127>
pe=00000 FAIL at line=00139    BaseServer.F90                           <status=-127>
pe=00000 FAIL at line=00986    ServerThread.F90                         <status=-127>
pe=00000 FAIL at line=00094    MessageVisitor.F90                       <status=-127>
pe=00000 FAIL at line=00113    AbstractMessage.F90                      <status=-127>
```
It doesn't go any farther than that. 

So I traced that last bit to `dispatch` and found some `dispatch` calls that weren't checking for return codes. I added some `_VERIFY(status)` there and then...still halts. 😦

**Comment by weiyuan-jiang** [comment]
Hopefully, the error messages get trapped now. @mathomp4

**Comment by mathomp4** [comment]
> Hopefully, the error messages get trapped now. @mathomp4

@weiyuan-jiang Well, it traps more, but like my test, your edits also deadlock:
```
 Writing:      5 Slices to File:  weiyuanmapl-gcm-2022Jan11-1day-c24-ChunkingFailure.geosgcm_budi.20000415_0000z.nc4
 Writing:     21 Slices to File:  weiyuanmapl-gcm-2022Jan11-1day-c24-ChunkingFailure.geosgcm_buda.20000414_2230z.nc4
pe=00000 FAIL at line=00701    NetCDF4_FileFormatter.F90                <status=-127>
pe=00000 FAIL at line=00297    NetCDF4_FileFormatter.F90                <status=-127>
pe=00000 FAIL at line=00063    HistoryCollection.F90                    <status=-127>
pe=00000 FAIL at line=00792    ServerThread.F90                         <status=-127>
pe=00000 FAIL at line=00138    BaseServer.F90                           <status=-127>
pe=00000 FAIL at line=00983    ServerThread.F90                         <status=-127>
pe=00000 FAIL at line=00094    MessageVisitor.F90                       <status=-127>
pe=00000 FAIL at line=00113    AbstractMessage.F90                      <status=-127>
```

It just stays there until you hit Ctrl-C.

**Comment by weiyuan-jiang** [comment]
I fixed the dispatch. If it is not working, then something else is wrong....

**Comment by mathomp4** [comment]
@weiyuan-jiang I pushed a compilation fix, but it still doesn't work. It now dies at:
```
pe=00000 FAIL at line=00107    SimpleSocket.F90                         <status=-127>
```

So, I tried going a bit farther. That is in a `send` routine, so I started adding `__RC__` to all the `%send` that looked viable.

That then stops at:
```
pe=00000 FAIL at line=00450    ClientThread.F90                         <status=-127>
```
That dies in `done_collective_stage`. So I then tried adding `__RC__` to all the `done_collective_stage` calls in pfio, and now:
```
 Writing:     21 Slices to File:  weiyuanmapl-gcm-2022Jan11-1day-c24-ChunkingFailure.geosgcm_buda.20000414_2230z.nc4
pe=00000 FAIL at line=00701    NetCDF4_FileFormatter.F90                <status=-127>
pe=00000 FAIL at line=00297    NetCDF4_FileFormatter.F90                <status=-127>
pe=00000 FAIL at line=00063    HistoryCollection.F90                    <status=-127>
pe=00000 FAIL at line=00798    ServerThread.F90                         <status=-127>
pe=00000 FAIL at line=00138    BaseServer.F90                           <status=-127>
pe=00000 FAIL at line=00989    ServerThread.F90                         <status=-127>
pe=00000 FAIL at line=00094    MessageVisitor.F90                       <status=-127>
pe=00000 FAIL at line=00113    AbstractMessage.F90                      <status=-127>
pe=00000 FAIL at line=00107    SimpleSocket.F90                         <status=-127>
pe=00000 FAIL at line=00450    ClientThread.F90                         <status=-127>
pe=00000 FAIL at line=00382    ClientManager.F90                        <status=-127>
pe=00000 FAIL at line=03634    MAPL_HistoryGridComp.F90                 <status=-127>
pe=00000 FAIL at line=01799    MAPL_Generic.F90                         <status=-127>
pe=00000 FAIL at line=01253    MAPL_CapGridComp.F90                     <status=-127>
pe=00000 FAIL at line=01182    MAPL_CapGridComp.F90                     <status=-127>
pe=00000 FAIL at line=01128    MAPL_CapGridComp.F90                     <status=-127>
pe=00000 FAIL at line=00796    MAPL_CapGridComp.F90                     <status=-127>
pe=00000 FAIL at line=00936    MAPL_CapGridComp.F90                     <status=-127>
pe=00000 FAIL at line=00261    MAPL_Cap.F90                             <status=-127>
pe=00000 FAIL at line=00218    MAPL_Cap.F90                             <status=-127>
pe=00000 FAIL at line=00152    MAPL_Cap.F90                             <status=-127>
pe=00000 FAIL at line=00129    MAPL_Cap.F90                             <status=-127>
pe=00000 FAIL at line=00034    GEOSgcm.F90                              <status=-127>
Abort(0) on node 0 (rank 0 in comm 0): application called MPI_Abort(MPI_COMM_WORLD, 0) - process 0
GEOSgcm Run Status: -1
```

🎉 

Would you like me to push my updates to this branch, or make a PR against your branch?

**Comment by weiyuan-jiang** [comment]
> 

Yes, please. It seems your branch finally traps the error.

**Comment by mathomp4** [comment]
> Yes, please. It seems your branch finally traps the error.

Done. I also hijacked the PR to fix up the CI a bit. (Even though #1288 is the future.)

Now to see what @tclune thinks. I *really* spread the `__RC__` (and therefore ESMF) all over. Maybe there's a way we can make an "ESMF-free" `__RC__`...

**Comment by mathomp4** [comment]
Oops. Now it crashes without a bad chunksize... Checking with @bena-nasa

**Comment by mathomp4** [comment]
Okay. @bena-nasa helped me fix the "good" case. He also had a suggestion about how to remove the new ESMF dependency by using `_RC` instead of `__RC__`. I'll try that now.

**Comment by tclune** [comment]
Ah - I'd forgotten there was a distinction.  But there is no good reason for `__RC__` to have a dependency on ESMF either.  ESMF cannot redefine true/false.  Just gives a convenient alias that we can skip.

**Comment by mathomp4** [comment]
> `__RC__` expands to
> 
> ```
> #define __RC__         RC=STATUS); _VERIFY(STATUS
> ```
> 
> It only induces a dependency on `MAPL_ErrorHandlingMod`.

I think the issue is that `__RC__` is only in `MAPL_Exceptions.h` and in `MAPL_Exceptions.h` we have:
https://github.com/GEOS-ESM/MAPL/blob/f98fae5c158838b991eeda51ebe3e7d93f4d6fea/include/MAPL_Exceptions.h#L157
which triggers the need (for include, not link).

But `_RC` is in `MAPL_ErrLog.h` which doesn't have the issue. My guess is we could do a "If ESMF not present during CMake, don't `#include "ESMC_ReturnCodes.h"` ?

**Comment by tclune** [comment]
Ah - I missed that.  We don't want a switch for this.  Just should evolve to avoid need for ESMF for error handling.  Do we even use the error codes?    ESMF provides no way to turn them into strings anyway.

**Comment by mathomp4** [comment]
> Ah - I missed that. We don't want a switch for this. Just should evolve to avoid need for ESMF for error handling. Do we even use the error codes? ESMF provides no way to turn them into strings anyway.

Huh. Well, it looks like we only look for ESMC codes in one file:
```console
$ rg esmc
base/MAPL_Config.F90
154:           _RETURN(ESMC_RC_LONG_STR)
162:           _RETURN(ESMC_RC_LONG_STR)
187:               _RETURN(ESMC_RC_LONG_STR)
195:               _RETURN(ESMC_RC_LONG_STR)
337:           _RETURN(ESMC_RC_LONG_STR)
345:           _RETURN(ESMC_RC_LONG_STR)
370:               _RETURN(ESMC_RC_LONG_STR)
378:               _RETURN(ESMC_RC_LONG_STR)
520:           _RETURN(ESMC_RC_LONG_STR)
528:           _RETURN(ESMC_RC_LONG_STR)
553:               _RETURN(ESMC_RC_LONG_STR)
561:               _RETURN(ESMC_RC_LONG_STR)
774:           _RETURN(ESMC_RC_LONG_STR)
782:           _RETURN(ESMC_RC_LONG_STR)
806:               _RETURN(ESMC_RC_LONG_STR)
814:               _RETURN(ESMC_RC_LONG_STR)

include/MAPL_Exceptions.h
157:#include "ESMC_ReturnCodes.h"
```
and `ESMF_RC` codes in a few others:
```console
$ rg esmf_rc
gridcomps/History/MAPL_HistoryGridComp.F90
2300:               if (status == ESMF_RC_FILE_READ) then

base/TeX/MAPL_ExceptionsDescr.tex
183:       case (ESMF_RC_CANNOT_GET)
185:       case (ESMF_RC_MEM)
230:   __raise__(ESMF_RC_ARG_RANK,'2D arrays are required here')
234:ESMF_RC_ARG_RANK: 2D arrays are required here

base/MAPL_Config.F90
115:      iret = ESMF_RC_NOT_IMPL
116:      if (present(rc)) rc = ESMF_RC_NOT_IMPL
129:        if ( iret .eq. ESMF_RC_NOT_FOUND ) then
226:        if ( iret .eq. ESMF_RC_NOT_FOUND ) iret = ESMF_SUCCESS
298:      iret = ESMF_RC_NOT_IMPL
299:      if (present(rc)) rc = ESMF_RC_NOT_IMPL
312:        if ( iret .eq. ESMF_RC_NOT_FOUND ) then
409:        if ( iret .eq. ESMF_RC_NOT_FOUND ) iret = ESMF_SUCCESS
481:      iret = ESMF_RC_NOT_IMPL
482:      if (present(rc)) rc = ESMF_RC_NOT_IMPL
495:        if ( iret .eq. ESMF_RC_NOT_FOUND ) then
592:        if ( iret .eq. ESMF_RC_NOT_FOUND ) iret = ESMF_SUCCESS
736:      iret = ESMF_RC_NOT_IMPL
737:      if (present(rc)) rc = ESMF_RC_NOT_IMPL
750:        if ( iret .eq. ESMF_RC_NOT_FOUND ) then
845:        if ( iret .eq. ESMF_RC_NOT_FOUND ) iret = ESMF_SUCCESS

generic/GenericCplComp.F90
1322:       RC = ESMF_RC_FILE_READ
```
Maybe we should just *explicitly* have the `include "ESMC_ReturnCodes.h"` line in the files that need them and remove from `MAPL_Exceptions.h`?

**Comment by tclune** [comment]
This is a separate issue.  Let's get this PR completed.

**Comment by mathomp4** [comment]
Re-blocking for one moment. I want to test with the multigroupserver too

**Comment by mathomp4** [comment]
Seems to work as hoped/expected for MultiGroup! Unblocking!

**Review by mathomp4 — APPROVED** [review]
Approving. 

I'll let @tclune take a look as well.

**Review by mathomp4 — APPROVED** [review]
Somehow dismissed my own review. Sigh.

File: griddedio/FieldBundleWrite.F90
Comment (mathomp4) [inline griddedio/FieldBundleWrite.F90:112]: @weiyuan-jiang @bena-nasa @tclune Should we add a:
```
_RETURN(_SUCCESS)
```
in this routine?

**Review by mathomp4 — APPROVED** [review]
Approve again

**Pull request merged**
